lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java

```diff
@@ -110,60 +129,90 @@ private HookResult processSingleHookDoc(Projector watchProjector, Projector incl
         return HookResult.aborted();
     }
 
-    private boolean watchedFieldsHaveChanged(JsonDoc preDoc, JsonDoc postDoc,
-            Projector watchProjector) {
+    /**
+     * Determines if any of the watched fields have changed. If
+     * arrayOrderingSignificant is true, also checks if array
+     * orderings also changed.
+     */
+    private boolean watchedFieldsHaveChanged(EntityMetadata metadata,
+                                             JsonDoc preDoc,
+                                             JsonDoc postDoc,
+                                             Projector watchProjector,
+                                             boolean arrayOrderingSignificant) {
         JsonDoc watchedPostDoc = watchProjector.project(postDoc, jsonNodeFactory);
         JsonDoc watchedPreDoc = preDoc == null
-                ? new JsonDoc(jsonNodeFactory.nullNode())
-                : watchProjector.project(preDoc, jsonNodeFactory);
+            ? new JsonDoc(jsonNodeFactory.objectNode())
+            : watchProjector.project(preDoc, jsonNodeFactory);
 
-        return !watchedPostDoc.getRoot().equals(watchedPreDoc.getRoot());
+        // Compute diff
+        JsonCompare cmp=metadata.getDocComparator();
+        try {
+            DocComparator.Difference diff=cmp.compareNodes(watchedPreDoc.getRoot(),watchedPostDoc.getRoot());
+            LOGGER.debug("Diff: {}",diff);
+            if(!diff.same()) {
+                if(diff.getNumChangedFields()>0) {
+                    return true;
+                } else {
+                    // Are there any array element moves?
+                    // Do we care?
+                    if(!diff.getDelta().isEmpty()&&arrayOrderingSignificant) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        } catch (Exception e) {
+            LOGGER.error("Error processing notification",e);
+            return false;
+        }
     }
 
     private NotificationEntity makeNotificationEntityWithIncludedFields(HookDoc hookDoc,
-            Projector includeProjector) {
+                                                                        Projector includeProjector) {
         EntityMetadata metadata = hookDoc.getEntityMetadata();
         JsonDoc postDoc = hookDoc.getPostDoc();
 
-        List<NotificationEntity.PathAndValue> data = new ArrayList<>();
-
+        NotificationEntity notificationEntity = new NotificationEntity();
+        // Set the payload
+        JsonDoc includeDoc=includeProjector.project(postDoc,jsonNodeFactory);
+        
+        List<NotificationEntity.PathAndValue> data=new ArrayList<>();
+        
+        JsonNodeCursor cursor=includeDoc.cursor();
+        while(cursor.next()) {
+            data.add(new NotificationEntity.PathAndValue(cursor.getCurrentPath().toString(),
+                                                         cursor.getCurrentNode().asText()));
```

Comment:
Good catch, we need to filter out container nodes.

On Thu, Jan 28, 2016 at 7:27 AM, Alec Henninger <notifications@github.com>
wrote:

> In
> lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java
> <https://github.com/esbtools/lightblue-notification-hook/pull/1#discussion_r51128295>
> :
>
> >          EntityMetadata metadata = hookDoc.getEntityMetadata();
> >          JsonDoc postDoc = hookDoc.getPostDoc();
> >
> > -        List<NotificationEntity.PathAndValue> data = new ArrayList<>();
> > -
> > +        NotificationEntity notificationEntity = new NotificationEntity();
> > +        // Set the payload
> > +        JsonDoc includeDoc=includeProjector.project(postDoc,jsonNodeFactory);
> > +
> > +        List<NotificationEntity.PathAndValue> data=new ArrayList<>();
> > +
> > +        JsonNodeCursor cursor=includeDoc.cursor();
> > +        while(cursor.next()) {
> > +            data.add(new NotificationEntity.PathAndValue(cursor.getCurrentPath().toString(),
> > +                                                         cursor.getCurrentNode().asText()));
>
> Does the cursor go through leaves only? (Do we need to filter out object
> nodes?)
>
> What does this do for array nodes?
>
> When we add tests, I'd expect to see scenarios for different field types
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/esbtools/lightblue-notification-hook/pull/1/files#r51128295>
> .
>